be/src/storage/vectorized/compaction.cpp

```diff
@@ -48,27 +104,59 @@ Status Compaction::do_compaction_impl() {
 
     // 1. prepare input and output parameters
     int64_t segments_num = 0;
+    int64_t total_row_size = 0;
     for (auto& rowset : _input_rowsets) {
         _input_rowsets_size += rowset->data_disk_size();
         _input_row_num += rowset->num_rows();
         segments_num += rowset->num_segments();
+        total_row_size += rowset->total_row_size();
     }
+
     TRACE_COUNTER_INCREMENT("input_rowsets_data_size", _input_rowsets_size);
     TRACE_COUNTER_INCREMENT("input_row_num", _input_row_num);
     TRACE_COUNTER_INCREMENT("input_segments_num", segments_num);
 
     _output_version = Version(_input_rowsets.front()->start_version(), _input_rowsets.back()->end_version());
 
+    // choose vertical or horizontal compaction algorithm
+    auto iterator_num_res = _get_segment_iterator_num();
```

Comment:
When we capture segment iterators, can we guarantee the segment_iterator_num is not larger than  MAX_SOURCES when using vertical algorithm?
Otherwise, the segment_iterator_num is larger than MAX_SOURCES will occurs OOM sometimes.

Reply:
```
max_cumulative_compaction_num_singleton_deltas=1000
base_compaction_num_cumulative_deltas=5
```
Depends on these two configurations, it would not be larger than MAX_SOURCES normally.